classes/war/Operation.php

```diff
@@ -328,16 +366,37 @@ public function handleCompletion(): string {
                     // update patrols, move back 5 minutes
                     $patrol_spawn = time() + (60 * 5);
                     $this->system->db->query("UPDATE `patrols` SET `start_time` = {$patrol_spawn}, `village_id` = {$this->user->village->village_id} WHERE `region_id` = {$location_target['region_id']}");
                     // update caravans, change only caravans that haven't spawned
                     $name = VillageManager::VILLAGE_NAMES[$this->user->village->village_id] . " Caravan";
                     $time = time();
-                    $this->system->db->query("UPDATE `caravans` SET `village_id` = {$this->user->village->village_id}, `name` = {$name} WHERE `region_id` = {$location_target['region_id']} AND `start_time` > {$time}");
+                    $this->system->db->query("UPDATE `caravans` SET `village_id` = {$this->user->village->village_id}, `name` = '{$name}' WHERE `region_id` = {$location_target['region_id']} AND `start_time` > {$time}");
+                    // for each occupied village in newly controlled region, if not at war then clear occupation
+                    $occupied_villages = $this->system->db->query("SELECT * FROM `region_locations` WHERE `region_id` = {$location_target['region_id']} AND `occupying_village_id` IS NOT NULL");
+                    $occupied_villages = $this->system->db->fetch_all($occupied_villages);
+                    if ($this->system->db->last_num_rows > 0) {
+                        foreach ($occupied_villages as $village) {
+                            if (!$this->user->village->isEnemy($village['occupying_village_id'])) {
+                                $this->system->db->query("UPDATE `region_locations` SET `occupying_village_id` = NULL WHERE `id` = {$village['id']}");
+                            }
+                        }
+                    }
+                } 
+                else if ($location_target['health'] <= 0 && $location_target['type'] == 'village') {
+                    WarLogManager::logAction($this->system, $this->user, 1, WarLogManager::WAR_LOG_VILLAGES_CAPTURED, $this->target_village);
+                    // occupy village and set HP/defense
+                    $village_hp = floor(WarManager::BASE_VILLAGE_HEALTH * 0.5);
+                    // if self or ally owns the region, set occupying village to null
+                    if ($this->user->village->isAlly($location_target['original_village'])) {
+                        $this->system->db->query("UPDATE `region_locations` SET `occupying_village_id` = NULL, `health` = {$village_hp}, `defense` = 25 WHERE `id` = {$location_target['id']}");
+                    } else {
+                        $this->system->db->query("UPDATE `region_locations` SET `occupying_village_id` = {$this->user->village->village_id}, `health` = {$village_hp}, `defense` = 25 WHERE `id` = {$location_target['id']}");
+                    }
```

Comment:
This is getting big, we could move to a `handleLocationDefeated` / `handleLocationCaptured` method

Reply:
Was thinking that, makes sense to me.